lib/Zonemaster/Backend/RPCAPI.pm

```diff
@@ -208,97 +251,105 @@ sub _check_domain {
-    }
+        if ( defined $syntax_input->{profile} ) {
+            my @profiles = map lc, Zonemaster::Backend::Config->load_config()->ListPublicProfiles();
+            return { status => 'nok', message => encode_entities( "Invalid profile option format" ) }
+            unless ( grep { $_ eq lc $syntax_input->{profile} } @profiles );
+        }
 
-    my ( $dn, $dn_syntax ) = $self->_check_domain( $syntax_input->{domain}, 'Domain name' );
+        my ( $dn, $dn_syntax ) = $self->_check_domain( $syntax_input->{domain}, 'Domain name' );
 
-    return $dn_syntax if ( $dn_syntax->{status} eq 'nok' );
+        return $dn_syntax if ( $dn_syntax->{status} eq 'nok' );
 
-    if ( defined $syntax_input->{nameservers} && @{ $syntax_input->{nameservers} } ) {
-        foreach my $ns_ip ( @{ $syntax_input->{nameservers} } ) {
-            my ( $ns, $ns_syntax ) = $self->_check_domain( $ns_ip->{ns}, "NS [$ns_ip->{ns}]" );
-            return $ns_syntax if ( $ns_syntax->{status} eq 'nok' );
-        }
+        if ( defined $syntax_input->{nameservers} && @{ $syntax_input->{nameservers} } ) {
+            foreach my $ns_ip ( @{ $syntax_input->{nameservers} } ) {
+                my ( $ns, $ns_syntax ) = $self->_check_domain( $ns_ip->{ns}, "NS [$ns_ip->{ns}]" );
+                return $ns_syntax if ( $ns_syntax->{status} eq 'nok' );
+            }
 
-        foreach my $ns_ip ( @{ $syntax_input->{nameservers} } ) {
-            return { status => 'nok', message => encode_entities( "Invalid IP address: [$ns_ip->{ip}]" ) }
-                unless( !$ns_ip->{ip} || $ns_ip->{ip} =~ /^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/ || $ns_ip->{ip} =~ /^([0-9A-Fa-f]{1,4}:[0-9A-Fa-f:]{1,}(:[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})?)|([0-9A-Fa-f]{1,4}::[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})$/);
+            foreach my $ns_ip ( @{ $syntax_input->{nameservers} } ) {
+                return { status => 'nok', message => encode_entities( "Invalid IP address: [$ns_ip->{ip}]" ) }
+                    unless( !$ns_ip->{ip} || $ns_ip->{ip} =~ /^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$/ || $ns_ip->{ip} =~ /^([0-9A-Fa-f]{1,4}:[0-9A-Fa-f:]{1,}(:[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})?)|([0-9A-Fa-f]{1,4}::[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3})$/);
 
-            return { status => 'nok', message => encode_entities( "Invalid IP address: [$ns_ip->{ip}]" ) }
-              unless ( !$ns_ip->{ip}
-                || Zonemaster::Engine::Net::IP::ip_is_ipv4( $ns_ip->{ip} )
-                || Zonemaster::Engine::Net::IP::ip_is_ipv6( $ns_ip->{ip} ) );
-        }
+                return { status => 'nok', message => encode_entities( "Invalid IP address: [$ns_ip->{ip}]" ) }
+                unless ( !$ns_ip->{ip}
+                    || Zonemaster::Engine::Net::IP::ip_is_ipv4( $ns_ip->{ip} )
+                    || Zonemaster::Engine::Net::IP::ip_is_ipv6( $ns_ip->{ip} ) );
```

Comment:
Aren't this two `return ... unless ...` statements doing exactly the same thing ?
I think the second one is a refactoring of the first one and therefore should be kept.

Reply:
One problem is that Net::IP accepts "network" as valid address. Therefore you must have two tests. The first test will catch network, e.g. "10.2.1", that the second would accept. The first one will accept "355.955.2.1", but that will be caught by the second test.

It looks odd so I guess some comment in the code is needed.

Reply:
Comment added in dcf35a6